zerver/tests/test_digest.py

```diff
@@ -140,6 +127,61 @@ def test_multiple_stream_senders(self,
         self.assertIn('some content', teaser_messages[0]['content'][0]['plain'])
         self.assertIn(teaser_messages[0]['sender'], expected_participants)
 
+    @mock.patch('zerver.lib.digest.enough_traffic')
+    @mock.patch('zerver.lib.digest.send_future_email')
+    def test_soft_deactivated_user_multiple_stream_senders(self,
+                                                           mock_send_future_email: mock.MagicMock,
+                                                           mock_enough_traffic: mock.MagicMock) -> None:
+
+        one_day_ago = timezone_now() - datetime.timedelta(days=1)
+        Message.objects.all().update(pub_date=one_day_ago)
+        one_sec_ago = timezone_now() - datetime.timedelta(seconds=1)
+
+        othello = self.example_user('othello')
+        for stream in ['Verona', 'Scotland', 'Denmark']:
+            self.subscribe(othello, stream)
+        RealmAuditLog.objects.all().delete()
+
+        othello.long_term_idle = True
+        othello.save(update_fields=['long_term_idle'])
+
+        # Send messages to a stream and unsubscribe - subscribe from that stream
+        senders = ['hamlet', 'cordelia',  'iago', 'prospero', 'ZOE']
+        self.simulate_stream_conversation('Verona', senders)
+        self.unsubscribe(othello, 'Verona')
+        self.subscribe(othello, 'Verona')
+
+        # Send messages to other streams
+        self.simulate_stream_conversation('Scotland', senders)
+        self.simulate_stream_conversation('Denmark', senders)
+
+        cutoff = time.mktime(one_sec_ago.timetuple())
+
+        flush_per_request_caches()
+        with queries_captured() as queries:
+            handle_digest_email(othello.id, cutoff)
+
+        self.assertTrue(12 <= len(queries) <= 13)
+
+        self.assertEqual(mock_send_future_email.call_count, 1)
+        kwargs = mock_send_future_email.call_args[1]
+        self.assertEqual(kwargs['to_user_ids'], [othello.id])
+
+        hot_conversations = kwargs['context']['hot_conversations']
+        self.assertEqual(2, len(hot_conversations), [othello.id])
+
+        hot_convo = hot_conversations[0]
+        expected_participants = {
+            self.example_user(sender).full_name
+            for sender in senders
+        }
+
+        self.assertEqual(set(hot_convo['participants']), expected_participants)
+        self.assertEqual(hot_convo['count'], 5 - 2)  # 5 messages, but 2 shown
+        teaser_messages = hot_convo['first_few_messages'][0]['senders']
+        self.assertIn('some content', teaser_messages[0]['content'][0]['plain'])
+        self.assertIn(teaser_messages[0]['sender'], expected_participants)
```

Comment:
Maybe we just need a comment explaining this, but I can't quite figure out how this test ensures that (1) the streams we want to filter get excluded and (2) the other ones don't.  

Possibly with the refactor I suggest above, this could be replaced with something that's more of a unit test checking the recipient IDs list selected, which would definitely have that property.

Reply:
Yes, that's indeed an insufficient test!